cpp/arcticdb/entity/timeseries_descriptor.hpp

```diff
@@ -12,73 +12,133 @@
 
 namespace arcticdb {
 
+struct FrameDescriptorImpl : public FrameDescriptor {
+    FrameDescriptorImpl() = default;
+
+    ARCTICDB_MOVE_COPY_DEFAULT(FrameDescriptorImpl)
+
+    [[nodiscard]] FrameDescriptorImpl clone() const {
+        return *this;
+    }
+};
+
 struct TimeseriesDescriptor {
-  using Proto = arcticdb::proto::descriptors::TimeSeriesDescriptor;
-
-  std::shared_ptr<Proto> proto_ = std::make_shared<Proto>();
-  std::shared_ptr<FieldCollection> fields_ = std::make_shared<FieldCollection>();
-  TimeseriesDescriptor() = default;
-
-  TimeseriesDescriptor(std::shared_ptr<Proto> proto, std::shared_ptr<FieldCollection> fields) :
-    proto_(std::move(proto)),
-    fields_(std::move(fields)) {
-  }
-
-  [[nodiscard]] std::shared_ptr<FieldCollection> fields_ptr() const  {
-      return fields_;
-  }
-
-  [[nodiscard]] std::shared_ptr<Proto> proto_ptr() const {
-      return proto_;
-  }
-
-  [[nodiscard]] bool proto_is_null() const {
-      return !proto_;
-  }
-
-  void set_stream_descriptor(const StreamDescriptor& desc) {
-      fields_ = std::make_shared<FieldCollection>(desc.fields().clone());
-      proto_ = std::make_shared<Proto>();
-      proto_->mutable_stream_descriptor()->CopyFrom(desc.proto());
-  }
-
-  [[nodiscard]] const FieldCollection& fields() const {
-      return *fields_;
-  }
-
-  [[nodiscard]] FieldCollection& mutable_fields() {
-      return *fields_;
-  }
-
-  [[nodiscard]] Proto& mutable_proto() {
-       return *proto_;
-  }
-
-  [[nodiscard]] const Proto& proto() const {
-      return *proto_;
-  }
-
-  [[nodiscard]] TimeseriesDescriptor clone() const {
-      auto proto = std::make_shared<Proto>();
-      proto->CopyFrom(*proto_);
-      return {std::move(proto), std::make_shared<FieldCollection>(fields_->clone())};
-  }
-
-  [[nodiscard]] StreamDescriptor as_stream_descriptor() const {
-      auto stream_descriptor = std::make_shared<arcticdb::proto::descriptors::StreamDescriptor>();
-      stream_descriptor->CopyFrom(proto_->stream_descriptor());
-      return StreamDescriptor(stream_descriptor, fields_);
-  }
-
-  void copy_to_self_proto() {
-      proto_->mutable_stream_descriptor()->mutable_fields()->Clear();
-      for(const auto& field : *fields_) {
-          auto new_field = proto_->mutable_stream_descriptor()->mutable_fields()->Add();
-          new_field->set_name(std::string(field.name()));
-          new_field->mutable_type_desc()->set_dimension(static_cast<uint32_t>(field.type().dimension()));
-          set_data_type(field.type().data_type(), *new_field->mutable_type_desc());
-      }
-  }
+    using Proto = arcticdb::proto::descriptors::FrameMetadata;
+
+    std::shared_ptr<FrameDescriptorImpl> frame_data_ = std::make_shared<FrameDescriptorImpl>();
+    std::shared_ptr<SegmentDescriptorImpl> segment_desc_ = std::make_shared<SegmentDescriptorImpl>();
+    std::shared_ptr<Proto> proto_ = std::make_shared<Proto>();
```

Comment:
It looks like we're still using protobuf for metadata. 

I assume that's fine because for a regular write/read we'll only read/write metadata just once, so should be insignificant compared to other encodings operations which will happen many times. So it's probably not worth migrating this complex type away from protobuf. Is that correct?

Reply:
User and Pandas metadata are in an opaque, optional field that's basically left for whatever implementers want to do with it. We'll have to support protobuf in there for the foreseeable future for backwards compatibility, but in the future where we're predominantly reading and writing Arrow data, we could maybe dump the pandas protobufs and store the user metadata directly in MsgPack or something.